design_notes.md

```diff
@@ -0,0 +1,228 @@
+# xoak Design notes
+
+## 1. Scope
+
+xoak aims at extending [xarray](https://github.com/pydata/xarray)'s capabilities
+to support efficient indexing of n-dimensional, irregular data using spatial
+partitioning tree structures.
+
+## 2. Spatial data partitioning trees
+
+Spatial data partitioning trees are commonly used as indexes for accelerating
+the following operations:
+
+- range search (e.g., find all points within given boundaries)
+- nearest neighbor lookup (e.g., find the point(s) in a given set that is/are
+  closest to a given point).
+
+There are many kinds of tree structures, e.g.,
+
+- k-d tree
+- ball tree
+- r-tree
+- quad-tree
+- octree
+- vantage-point tree
+- ...
+
+Each structure has its own advantages and limitations regarding index
+construction time, query time, memory footprint and types of search supported,
+that mostly depend on the properties of the space being partitioned, e.g., low
+vs high dimensions, euclidean vs non-euclidean, etc.
+
+## 3. Reuse vs. implement tree structures
+
+A few structures are available in Python:
+
+- A kd-tree is implemented in the
+  [scipy.spatial](https://docs.scipy.org/doc/scipy/reference/spatial.html)
+  sub-package
+- [libpysal](https://github.com/pysal/libpysal) adds support for arc distance on
+  top of scipy's k-d tree
+- [scikit-learn](https://scikit-learn.org) has k-d tree and ball tree implementations
+- The [rtree](https://github.com/Toblerity/rtree) package wraps
+  [libspatialindex](https://github.com/libspatialindex/libspatialindex)
+
+xoak will reuse (some of) those existing implementations. In the mid/long term,
+however, xoak might benefit of its own implementations. Motivations are:
+
+- Data structures not available elsewhere (Python packages)
+- More flexible implementations (e.g., allow user-defined distance functions)
+- More control of underlying parallelisation, e.g., with Dask
+
+Possible solutions for efficient implementation and their main advantage:
+
+- Cython: already used by scikit-learn and scipy
+- Numba: mature enough and flexible (would allow user-defined metrics
+  implemented as jitted functions)
+- C++ / Xtensor: reuse the powerful features of the C++ standard library /
+  ecosystem.
+
+## 4. API
+
+### xarray Dataset / DataArray accessor(s)
+
+The current, recommended way to extend Xarray is via
+[accessors](http://xarray.pydata.org/en/stable/internals.html#extending-xarray).
+
+#### One vs. multiple accessors
+
+We have two options:
+
+A. one accessor per kind of index, e.g., ``Dataset.kdtree``, ``Dataset.balltree``, etc.
+B. one unique accessor ``Dataset.xoak`` for handling all kinds of indexes.
```

Comment:
Formatting issue (at least in the preview this would otherwise all be in one line):
```suggestion
- *A.* one accessor per kind of index, e.g., ``Dataset.kdtree``, ``Dataset.balltree``, etc.
- *B.* one unique accessor ``Dataset.xoak`` for handling all kinds of indexes.
```